Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h

```diff
@@ -113,6 +114,7 @@ class TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLayer> {
     bool syncAnimations(MonotonicTime);
     WEBCORE_EXPORT bool descendantsOrSelfHaveRunningAnimations() const;
 
+    WEBCORE_EXPORT void prepareForPainting(TextureMapper&);
```

Comment:
I don't understand why you have to change the public API. After this change, all client have to call `prepareForPainting`. Why not `paint` calls `prepareForPainting` internally?

Reply:
Good point, I can call `prepareForPainting` in `paint` iff `!ENABLE(DAMAGE_TRACKING)`

Reply:
I don't understand why it's only for `!ENABLE(DAMAGE_TRACKING)`. 
Does `beginClip` have to be called between `beginPainting` and `prepareForPainting` if ENABLE(DAMAGE_TRACKING)?

Reply:
Well, in case of `ENABLE(DAMAGE_TRACKING)` the mandatory sequences are:

1. `beginPainting` -> `beginClip` -> `collectDamage`
2. `prepareForPainting` -> `collectDamage`

In the above:

- `beginPainting` and `beginClip` set the clip region to the latest viewport size; therefore when `collectDamage` is invoked, it can trim damage rectangles to the clip region.
- `prepareForPainting` calculates transforms of all layers recursively; therefore when `collectDamage` is invoked, it can position damage rectangles correctly (in the viewport coordinate space)

In case of `!ENABLE(DAMAGE_TRACKING)` everything can be done as it was before.

Reply:
Got it.